components/ee/ws-scheduler/pkg/scheduler/scheduler.go

```diff
@@ -350,6 +350,27 @@ func (s *Scheduler) buildState(ctx context.Context, pod *corev1.Pod) (state *Sta
 	// Don't forget to take into account the bindings we just created (e.g., are still in the cache)
 	// to avoid assigning slots multiple times
 	state.UpdateBindings(s.localBindingCache.getListOfBindings())
+
+	// The required node services is basically PodAffinity light. They limit the nodes we can schedule
+	// workspace pods to based on other pods running on that node. We do this because we require that
+	// ws-daemon and registry-facade run on the node.
+	//
+	// Alternatively, we could have implemented PodAffinity in ws-scheduler, but that's conceptually
+	// much heavier and more difficult to handle.
+	//
+	// TODO(cw): if we ever implement PodAffinity, use that instead of requiredNodeServices.
+	if rs, ok := pod.Annotations[wsk8s.RequiredNodeServicesAnnotation]; ok {
```

Comment:
Conceptually I'd expect an additional filter before `state.UpdateNodes(potentialNodes)`, because it trims down the list of potential candidates.

Reply:
I'm not sure we can do that, because only in the state do we establish the connection between pods and nodes. We need that connection to implement this filter.